### ISSUE TITLE
3857 to 4326 issue

### ISSUE THREAD

**Issue by moovida** [maRci002/proj4dart#1]
Hi, I am sure I am doing something wrong here, but I can't find what and how, so after hours of trial, I will ask you to see if there is an issue. :-)

I am converting a 3857 bbox to 4326.

```
      var originPrj = proj4dart.Projection.parse(wktPrj);
      var destPrj = proj4dart.Projection("EPSG:4326");

      var ll = proj4dart.Point(x: west, y: south);
      var llDest = originPrj.transform(destPrj, ll);
```
where wktPrj is read from a prj file. That one is read properly and recognises the projection correctly (PseudoMercatorProjection).

But the converted Y for some reason is off. 
For example converting:
west: 1235116.6370151439
south: 5852571.412104731

should give me:
x: 11.0952417
y: 46.4488475

but the y I get is:
y: 46.6408787

Do you have any idea what that might happen?

**Comment by maRci002** [comment]
Hi, @moovida thanks for the first issue :tada:

This one seems bit tricky beacuse even proj4js does a `checkMercator` [method call after pasing the WKT string](https://github.com/proj4js/proj4js/blob/master/lib/parseCode.js#L42-L47) and since [3857](https://epsg.io/3857) is a Mercator projection they throw away parsed WKT and return with cached one which will give the right results.

If I remove this code from proj4js then their result will be: `[11.095241527010067, 46.448847417864975]` which still still right.
```
      // test of spetial case, due to this being a very common and often malformed
      if (checkMercator(out)) {
        return defs['EPSG:3857'];
      }
```

I think I could create our own `checkMercator` method and return our cached one `Projection('EPSG:3857')` which was parsed from [proj4 string](https://epsg.io/3857.proj4) which gives the right result: `x: 11.095241527010067, y: 46.448847417864975` however it annoys me they do the right calculations even without `checkMercator`.

I will investigate this issue a little bit later.
Until new version please use  this as a workaround `var originPrj = Projection('EPSG:3857');` if wkt string equals to [this](//epsg.io/3857.wkt).

**Comment by moovida** [comment]
Hi @maRci002 , thank you for this fantastic project.

I am not sure I understand why the result is wrong. You are telling me that it doesn't recognise the 3857 from my WKT prj, right?

BTW, my wkt is the following:
```
PROJCS["WGS 84 / Pseudo-Mercator", 
  GEOGCS["WGS 84", 
    DATUM["World Geodetic System 1984", 
      SPHEROID["WGS 84", 6378137.0, 298.257223563, AUTHORITY["EPSG","7030"]], 
      AUTHORITY["EPSG","6326"]], 
    PRIMEM["Greenwich", 0.0, AUTHORITY["EPSG","8901"]], 
    UNIT["degree", 0.017453292519943295], 
    AXIS["Geodetic latitude", NORTH], 
    AXIS["Geodetic longitude", EAST], 
    AUTHORITY["EPSG","4326"]], 
  PROJECTION["Popular Visualisation Pseudo Mercator", AUTHORITY["EPSG","1024"]], 
  PARAMETER["semi_minor", 6378137.0], 
  PARAMETER["latitude_of_origin", 0.0], 
  PARAMETER["central_meridian", 0.0], 
  PARAMETER["scale_factor", 1.0], 
  PARAMETER["false_easting", 0.0], 
  PARAMETER["false_northing", 0.0], 
  UNIT["m", 1.0], 
  AXIS["Easting", EAST], 
  AXIS["Northing", NORTH], 
  AUTHORITY["EPSG","3857"]]
```

But while debugging it seemed to be parsing it right, so I don't understand what is going on. Given the parameters contained in the WKT and in the https://epsg.io/3857.proj4 I would expect the reprojection to be done right in both cases? Sorry, I am a bit confused.

**Comment by maRci002** [comment]
WKT and proj4 definitions may produce different results or even using different Porjection algorithm under the hood.
A good example is https://epsg.io/23700

- [If using proj4 def](https://epsg.io/23700.proj4): then it will use `somerc` / `SwissObliqueMercatorProjection` <= `+proj=somerc `
- [If using OGC WKT def](https://epsg.io/23700.wkt): then it will use `omerc` / `HotineObliqueMercatorProjection` <= `,PROJECTION["Hotine_Oblique_Mercator_Azimuth_Center"]`
- [If using ESRI WKT def](https://epsg.io/23700.esriwkt): then it will use `omerc` / `HotineObliqueMercatorProjection` <= `,PROJECTION["Hotine_Oblique_Mercator_Azimuth_Center"]`

All of them will reproduce different results (even when `esri wkt` / `ogc wkt` using same projection)

**Check out proj4js results:**
```
// proj4 def
var firstProjection = '+proj=somerc +lat_0=47.14439372222222 +lon_0=19.04857177777778 +k_0=0.99993 +x_0=650000 +y_0=200000 +ellps=GRS67 +towgs84=52.17,-71.82,-14.9,0,0,0,0 +units=m +no_defs ';
proj4(firstProjection,"WGS84",[561651.8408065987, 172658.61998377228]);
// prints [17.88805856557485, 46.89226406698969]
```

```
// ogc wkt def
var firstProjection = 'PROJCS["HD72 / EOV",GEOGCS["HD72",DATUM["Hungarian_Datum_1972",SPHEROID["GRS 1967",6378160,298.247167427,AUTHORITY["EPSG","7036"]],TOWGS84[52.17,-71.82,-14.9,0,0,0,0],AUTHORITY["EPSG","6237"]],PRIMEM["Greenwich",0,AUTHORITY["EPSG","8901"]],UNIT["degree",0.0174532925199433,AUTHORITY["EPSG","9122"]],AUTHORITY["EPSG","4237"]],PROJECTION["Hotine_Oblique_Mercator_Azimuth_Center"],PARAMETER["latitude_of_center",47.14439372222222],PARAMETER["longitude_of_center",19.04857177777778],PARAMETER["azimuth",90],PARAMETER["rectified_grid_angle",90],PARAMETER["scale_factor",0.99993],PARAMETER["false_easting",650000],PARAMETER["false_northing",200000],UNIT["metre",1,AUTHORITY["EPSG","9001"]],AXIS["Y",EAST],AXIS["X",NORTH],AUTHORITY["EPSG","23700"]]';
proj4(firstProjection,"WGS84",[561651.8408065987, 172658.61998377228]);
// prints [17.88805977211558, 46.89226406836302]
```

```
// esri wkt def
var firstProjection = 'PROJCS["HD72_EOV",GEOGCS["GCS_HD72",DATUM["D_Hungarian_1972",SPHEROID["GRS_1967",6378160,298.247167427]],PRIMEM["Greenwich",0],UNIT["Degree",0.017453292519943295]],PROJECTION["Hotine_Oblique_Mercator_Azimuth_Center"],PARAMETER["latitude_of_center",47.14439372222222],PARAMETER["longitude_of_center",19.04857177777778],PARAMETER["azimuth",90],PARAMETER["scale_factor",0.99993],PARAMETER["false_easting",650000],PARAMETER["false_northing",200000],UNIT["Meter",1]]';
proj4(firstProjection,"WGS84",[561651.8408065987, 172658.61998377228]);
// prints [17.88916689817367, 46.89253683075668]
```

I think the proj4 def will give the best result.

Here is more results from proj4js (almost every forward / inverse will give different results):
- proj4 def results: [all_proj4_results.dart](https://github.com/maRci002/proj4dart/blob/master/test/results/all_proj4_results.dart)
- ogc wkt def results: [all_proj4_results.dart](https://github.com/maRci002/proj4dart/blob/master/test/results/all_proj4_ogc_wkt_results.dart)
- esri wkt def results: [all_proj4_results.dart](https://github.com/maRci002/proj4dart/blob/master/test/results/all_proj4_esri_wkt_results.dart)

**Comment by moovida** [comment]
Thanks a ton for the explanation @maRci002 , much appreciated!

**Comment by fegyi001** [comment]
Hello everyone, I'm also a contributor of proj4dart. Just recently checked this issue, and I hereby confirm, that our code malfunctions. The WKT parsing part works well, but after that something goes wrong which still needs to be examined. 

What I did: I tested how WKT gets parsed. Proj4dart uses wkt-parser under the hood which is the dart porting of proj4js/wkt-parser. I examined the output created by both the dart and the javascript version (the javascript can be regarded as the "original"), and they produce the exact same results. I tested it with 3 different definitions for EPSG:3857 (the OGC WKT definition from epsg.io, the ESRI WKT definition from epsg.io and the one that @moovida provided), in all 3 cases the json output is exactly the same. So the wkt parsing part of the process is right.

I also tested proj4js with the given input coordinated @moovida uses, and I saw that the javascript version of proj4 always results the same latlon coordinates while in case of proj4dart only the ESRI WKT version returns correct values, the OGC WKT and the def from @moovida returns that strange second coordinates.

**Comment by fegyi001** [comment]
Well, this starts to get interesting. It seems that proj4js does not work well either, but it makes 2 workarounds to return the proj4 string definition of EPSG:3857. 

1. check whether the parsed WKT definition is EPSG:3857 or anything that is similar to it. But this is only to make sure that if the programmer mistyped the epsg code he/she still could use the correct one. The "testMercator" method comment is: "test of spetial case, due to this being a very common and often malformed". I guess whe should implement this as well.

2. This is quite a hack from proj4js: "maybeProjStr" method. It checks whether the WKT definition has an "EXTENSION" property, which contains a proj4 string definition instide a WKT definition. This is weird in itself, but in such cases proj4js simply uses this proj4 string definition INSTEAD the whole WKT definition. 

If I comment out this two checks, proj4js also returns [11.095241527010067, 46.64087873319058], the same proj4dart returns as well. Wow, I did not expect that.

So this means if we also implement these two hacky-smell methods this would solve it.

**Comment by maRci002** [comment]
Thanks to @fegyi001 it is fixed in `v1.0.5` version!
